src/util/sockopt.c

```diff
@@ -194,3 +195,44 @@ int sockopt_get_peergroups(int fd, Log *log, uid_t uid, gid_t primary_gid, gid_t
 
         return 0;
 }
+
+int sockopt_get_peerpidfd(int fd, pid_t pid, int *ret_pidfd) {
+#ifdef SO_PEERPIDFD
+        socklen_t socklen = sizeof(int);
+#endif
+        int r, pidfd = -1;
+
+        assert(fd >= 0);
+        assert(ret_pidfd);
+
+        /*
+         * New option from kernel 6.4. We'll return this with credentials,
+         * so that access policies that are resistent from PID reuse attacks
+         * can be implemented. Return '1' if SO_PEERPIDFD is supported and works.
+         * If we don't have SO_PEERPIDFD, we'll fall back to the next best
+         * thing, which is to pin the pid we got from SO_PEERCRED with a pidfd.
+         * This doesn't completely close the race window, but it makes it
+         * shorter as we do this immediately after receiving the creds. In this
+         * case we return '0', to signal to the caller that the pidfd cannot be
+         * completely trusted.
+         */
```

Comment:
Yes, this won't be available in a LTS kernel till the end of the year assuming everything goes well, and for embedded devices realistically it will be at least a couple years until the new kernel option is available anywhere. Userspace updates will land much more quickly than that.